### PR TITLE
Custom Validation

### PR DESCRIPTION
**Pull request by dpatil96**

### CODE REVIEW

File: app/models/cart_item.rb [hunk 1]
```diff
@@ -5,4 +5,22 @@ class CartItem < ApplicationRecord
   belongs_to :product
   validates :price, presence: true
   belongs_to :discount, optional: true
-end
+  validate :check_age_restriction
+
+
+  def check_age_restriction
+    if product.age_restricted? && user&.profile&.age.to_i  < 18
```
Comment (vtekiosi) [inline app/models/cart_item.rb:12]: user&.profile ? every user instance will have profile for sure. right?

File: app/controllers/cart_items_controller.rb [hunk 2]
```diff
@@ -6,7 +6,9 @@ class CartItemsController < ApplicationController
   before_action :set_cart_item, only: %i[update destroy]
 
   def create
-    if @cart.save
+    @cart_item = current_user.cart.cart_items.new(cart_item_params)
+
+    if @cart_item.check_age_restriction && @cart_item.save
```
Comment (vtekiosi) [inline app/controllers/cart_items_controller.rb:11]: check_age_restriction method not required to call from here. the save method automatically triggers validations. check the model callbacks for reference

File: app/controllers/cart_items_controller.rb [hunk 3]
```diff
@@ -47,32 +49,6 @@ def set_cart_item
   end
 
   def cart_item_params
-    params.require(:cart_item).permit(:quantity)
+    params.require(:cart_item).permit(:product_id, :quantity, :price, :user_id)
```
Comment (vtekiosi) [inline app/controllers/cart_items_controller.rb:52]: does cart_item require user_id attribute here? we already have the cart linked to user. right?

File: app/controllers/custom_sessions_controller.rb [hunk 4]
```diff
@@ -1,7 +1,13 @@
 # frozen_string_literal: true
 
-class SessionsController < ApplicationController
-  def new; end
+class CustomSessionsController < Devise::SessionsController
+  
+  def create
+    super do |user|
+      session[:user_id] = user.id
```
Comment (vtekiosi) [inline app/controllers/custom_sessions_controller.rb:7]: block missing end here

File: app/controllers/profiles_controller.rb [hunk 5]
```diff
@@ -6,7 +6,7 @@ class ProfilesController < ApplicationController
   def edit; end
 
   def update
-    if @profile.update(profile_params)
+    if @profile.update(profile_params)
```
Comment (vtekiosi) [inline app/controllers/profiles_controller.rb:9]: remove additional spacing

File: app/models/profile.rb [hunk 6]
```diff
@@ -2,4 +2,17 @@
 
 class Profile < ApplicationRecord
   belongs_to :user
+  after_update :set_age
+
+  def set_age
+    if birth_date.present?
+      today = Date.today
+      calculated_age = today.year - birth_date.year
+      calculated_age -= 1 if today < birth_date + calculated_age.years
```
Comment (vtekiosi) [inline app/models/profile.rb:11]: birth date cant be a future value. please refactor the code

**Review by vtekiosi — COMMENTED** [review]
please look into review comments

File: app/models/cart_item.rb (hunk 1 above)
Comment (dpatil96) [reply app/models/cart_item.rb:12]: yes every user will have a profile.I made a associations between user and profile.

**Pull request merged**